### PR TITLE
Complex navigator display fixing

### PR DESCRIPTION
**Pull request by jmedinaebi**

### CODE REVIEW

File: src/app/basket/basket.component.css [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.ComplexNavigator {
+  float: left;
+  width: 100%
+  /*background-color: #6dab49;*/
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.css]: Remove this

File: src/app/basket/basket.component.ts [hunk 2]
```diff
@@ -11,13 +15,18 @@ import {Title} from '@angular/platform-browser';
 })
 export class BasketComponent implements OnInit, AfterViewInit {
   private _complexBasket: { [name: string]: BasketItem };
+  private _query: string;
+  complexSearchBasket: ComplexSearchResult = null;
+  allInteractorsInComplexSearchBasket: Interactor[] = [];
+  private _complexes: Element[];
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.ts]: Not used, remove.

File: src/app/basket/basket.component.ts [hunk 3]
```diff
@@ -45,4 +55,51 @@ export class BasketComponent implements OnInit, AfterViewInit {
...
+    });
+  }
+
+  complexNavigatorLoading() {
+    this._query = this.createQuery(this._complexBasket);
+    this.complexSearchBasket = null;
+    this.allInteractorsInComplexSearchBasket = [];
+    this._complexes = [];
+    this.requestComplexesForNavigator();
+    console.log(this._complexes);
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.ts]: Remove console.log calls

File: src/app/basket/basket.component.html [hunk 4]
```diff
@@ -13,13 +13,13 @@ <h1>Personal complex collection</h1>
       <tbody>
       <tr *ngFor="let key of getKeys(complexBasket)">
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.html:28]: Should we get rid of the list? Or add list/navigator buttons, similar to the ones in the search?

File: src/app/complex/complex-details/complex-participants/complex-participant-legend/complex-participant-legend.component.html [hunk 5]
```diff
@@ -1,58 +1,58 @@
...
+        {{ getConvertedStochiometry(participant.stochiometry) }}</p>
+    </div>
+  </td>
+</tr>
+<ng-container *ngIf="participant.participants.length > 0">
+  <tr></tr>
+  <tr class="text-center small-row" *ngIf="participant.participants.length > 0">
+    <td style="{{getBackgroundStyle()}}"></td>
+    <td colspan="2" *ngIf="participant.expanded" class="no-padding-top">
+      <a class="label" (click)="participant.expanded = false">Hide components</a>
```
Comment (jmedinaebi) [inline src/app/complex/complex-details/complex-participants/complex-participant-legend/complex-participant-legend.component.html:36]: Should we keep this as subcomponents?

File: src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css [hunk 6]
```diff
@@ -16,7 +16,6 @@
 }
 
 .filters {
-  width: 120.5%;
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css:19]: Why was this removed?

File: src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css [hunk 7]
```diff
@@ -33,6 +32,10 @@
   display: none;
 }
 
+label {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css:37]: Why do we need this?

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts [hunk 8]
```diff
@@ -204,59 +208,107 @@ export class TableInteractorColumnComponent implements OnChanges {
...
+
+  private calculateTimesAppearingType() {
+    this._timesAppearingByType = new Map();
     for (const oneInteractor of this.enrichedInteractors) {
       for (const complex of this.complexes) {
         const match = findInteractorInComplex(complex, oneInteractor.interactor.identifier, this.enrichedInteractors);
         if (!!match) {
-          if (!!match.stochiometryValue) {
-            oneInteractor.timesAppearing += match.stochiometryValue[0];
+          oneInteractor.timesAppearing += 1;
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts]: `oneInteractor.timesAppearing += 1;` is called in this method and the method below, so we're setting `timesAppearing` to double the real value.
 It doesn't matter much, as all interactors have their number duplicated, so they are still in order, but to fix it we have 2 options.

1. Merge this 2 methods into 1, that does it all. They both have the same for loops, so it will just moving the if/else from one method to another.

2. Keep these 2 methods, without the `oneInteractor.timesAppearing += 1;` line, and have another method, `calculateTimesAppearing()` that only does that. Then call the 3 method on the `onChanges` method.

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.css [hunk 9]
```diff
@@ -1,11 +1,26 @@
+:host {
+  display: flex;
+  flex-direction: column;
+  height: 100%;
+  justify-content: center;
+  align-items: center;
+  align-content: center;
+  flex-wrap: nowrap;
+
+  * {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.css:10]: What's `*`?

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 10]
```diff
@@ -20,7 +20,7 @@ export class TableStructureComponent implements OnChanges {
 
   ngOnChanges(): void {
     this.classifyComplexesSize(); // ensure to always have the same base of comparison
-    this.sortedComplexes = this.classifyComplexesSimilarities(this.complexSearch.elements);
+    this.sortedComplexes = this.classifyComplexesSimilaritiesV2(this.complexSearch.elements);
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:26]: What has changed here?

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html [hunk 11]
```diff
@@ -19,26 +19,19 @@
     </a>
 
     <div class="nameCompact" *ngIf="!IDDisplay">
-      <div class="name" title="More complexes containing {{interactorName}}" *ngIf="!IDDisplay">
-        <a *ngIf="!IDDisplay"
-           [routerLink]="['/complex/search']"
-           [queryParams]="{query: interactorId, page: 1}"
-           target="_blank">
+      <div class="name" title="More information about containing {{interactorName}}" *ngIf="!IDDisplay">
+        <a href="{{identifierLink}}" target="_blank" class="externalLinkContainer">
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html]: Add `*ngIf="!!identifierLink"` to this `<a>` element.

**Comment by jmedinaebi** [comment]
I don't know if the code in the github pages is up to date, but I just tried and the search results look wonky.

For https://complex-portal.github.io/complex-portal-view/complex/search?query=%22nuclear%20pore%22#view_list

Screenshot 2024-06-20 at 13 25 30
Screenshot 2024-06-20 at 13 25 36

**Comment by jmedinaebi** [comment]
I see a lot of CSS changes, are those for the Safari and Firefox support?

File: src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.html [hunk 12]
```diff
@@ -2,12 +2,12 @@
   <div class="sortingInteractors">
     <input type="submit" name="classificationButton" value="Sort interactors" class="button">
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.html]: Sort components

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html [hunk 13]
```diff
@@ -28,7 +28,7 @@
           <!-- Icon for the expandable -->
           <div class="container" *ngIf="interactor.isSubComplex">
             <a title="{{ interactor.interactor.name }} interactors" (click)="toggleSubcomplexExpandable(i)">
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html]: `<a title="{{ interactor.interactor.name }} interactors"...` should be `<a title="{{ interactor.interactor.name }} components" ...`

Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html]: Same in lines 58 and 61

File: src/app/complex/complex-details/complex-participants/complex-participants.component.html [hunk 14]
```diff
@@ -1,5 +1,5 @@
 <div class="row" *ngIf="participantsWithSubcomponents">
-  <h2 class="float-left">Participants</h2>
+  <h2 class="float-left">Components</h2>
```
Comment (jmedinaebi) [inline src/app/complex/complex-details/complex-participants/complex-participants.component.html:2]: Since we change this, we should also change it in `go-to.component.html`, line 9, from Participants to Components.

Comment (jmedinaebi) [reply src/app/complex/complex-details/complex-participants/complex-participants.component.html:2]: Then, in `complex-filter.component.html`, should it be `Interactor Type` or `Component Type`? Can you check that with Sucharita?

File: src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css (hunk 7 above)
Comment (susiehgt) [reply src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css:37]: We made some modification for the menus so the user can click directly on the whole tile rectangle instead of clicking on the radiobutton, to do so we encapsulated it in the label which needed some style adjustments

File: src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css (hunk 6 above)
Comment (susiehgt) [reply src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.css:19]: As we moved the complex navigator into a div to be modified depending on the basket and the search result page, the buttons were not centered anymore with this width

File: src/app/complex/complex-details/complex-participants/complex-participants.component.html (hunk 14 above)
Comment (jmedinaebi) [reply src/app/complex/complex-details/complex-participants/complex-participants.component.html:2]: As Sucharita said, let's rename this as Component Type.

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.css (hunk 9 above)
Comment (susiehgt) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.css:10]: it selects all the elements contained in :host to apply the flex on them at the same rate, it's there for the firefox and safari compatibility

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts (hunk 10 above)
Comment (susiehgt) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:26]: the classifyComplexesSimilaritiesV2 is the function supposed to cluster the complexes but it is not finished yet, right now it is only classifying the complexes depending on the similarity score they obtain

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts [hunk 15]
```diff
@@ -327,4 +378,19 @@ export class TableInteractorColumnComponent implements OnChanges {
...
+  interactorsTimesAppearing() {
+    for (const oneInteractor of this.enrichedInteractors) {
+      for (const complex of this.complexes) {
+        const match = findInteractorInComplex(complex, oneInteractor.interactor.identifier, this.enrichedInteractors);
+        if (!!match) {
+          if (!!match.stochiometryValue) {
+            oneInteractor.timesAppearing += match.stochiometryValue[0];
+          } else {
+            oneInteractor.timesAppearing += 1;
+          }
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts]: We could simply replace these 5 lines with
```
oneInteractor.timesAppearing += 1;
```

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 16]
```diff
@@ -100,14 +102,16 @@ export class TableStructureComponent implements OnChanges {
...
         }
       }
+      similarityArray[2] = similarities;
     }
+    // console.log(similarityArray);
     return similarities;
   }
 
-  private classifyComplexesSimilarities(bigComplexes: Element[]): Element[] {
+  private classifyComplexesSimilarities(complexesList: Element[]): Element[] {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts]: We can ger rid of this method, right?

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 17]
```diff
@@ -132,4 +136,59 @@ export class TableStructureComponent implements OnChanges {
...
+
+    const complexesOrderedSet = this.uniqueComplexesListOrderedBySimilarity(comparedComplexes);
+    // to be used in the table as a 1D array
+    for (const complex of complexesOrderedSet) {
+      finalList.push(complex);
+    }
+    return finalList;
+  }
+
+  uniqueComplexesListOrderedBySimilarity(complexesListSimilarities: [Element, Element, number][]) {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:77]: We have 2 for loops here where we iterate through the complexesListSimilarities list. It makes sense but it can be more slightly more efficient.

The second time we iterate we can start from where we are, as anything before that point has already been checked. To do that, we can use indexes in the for loop.

```
    for (let i = 0; i < complexesListSimilarities.length; i++) {
      const complex1 = complexesListSimilarities[i][0];
      const complex2 = complexesListSimilarities[i][1];
      const similarityScore = complexesListSimilarities[i][2];
      if (similarityScore !== 0) {
        if (!complexesOrderedSet.has(complex1)) {
          complexesOrderedSet.add(complex1);
        }
        if (!complexesOrderedSet.has(complex2)) {
          complexesOrderedSet.add(complex2);
        }
        for (let j = i + 1; j < complexesListSimilarities.length; j++) {
          const complex3 = complexesListSimilarities[j][0];
          const complex4 = complexesListSimilarities[j][1];
          const similarityScore2nd = complexesListSimilarities[j][2];
          if (complex1 === complex3 && similarityScore2nd !== 0) {
            if (!complexesOrderedSet.has(complex4)) {
              complexesOrderedSet.add(complex4);
            }
          }
        }
      }
    }
```

I think that would make this from O(n^2) to O(nLogn), and it should work the same way.

**Comment by jmedinaebi** [comment]
You have changed the interactor(s)/participant(s) labels to component(s) everywhere but `go-to.component.html`, which you haven't modified, so it makes sense anyway. Can you replace Participants with Components there in line 9?

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts (hunk 17 above)
Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:77]: Basically, if we have 4 complexes (CPX1, CPX2, CPX3 and CPX4), `complexesListSimilarities` could look like
```
complexesListSimilarities = [
  [CPX1, CPX1, 5],
  [CPX2, CPX2, 5],
  [CPX1, CPX2, 4],
  [CPX3, CPX3, 3],
  [CPX1, CPX4, 2],
  [CPX4, CPX4, 2],
  [CPX2, CPX3, 1],
  [CPX2, CPX4, 1],
  [CPX3, CPX4, 0],
  [CPX1, CPX3, 0]
]
```

First, on the first loop, with i=0, we check `[CPX1, CPX1, 5]` and we add CPX1 to the set.
Then, we do the second loop from j=1 (i+1), to j=length-1, and we add CPX2 and CPX4 to the set.

Then, i=1, CPX2 is already added to the set. The second loop now starts from j=2.
CPX3 is added to the set.

The rest of iterations on the loops will not add anything, so we end up with [CPX1, CPX2, CPX4, CPX3].

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 18]
```diff
@@ -132,4 +136,59 @@ export class TableStructureComponent implements OnChanges {
...
+        if (!complexesOrderedSet.has(complex2)) {
+          complexesOrderedSet.add(complex2);
+        }
+        for (const [complex3, complex4, similarityScore2nd] of complexesListSimilarities) {
+          if (complex1 === complex3 && similarityScore2nd !== 0) {
+            // console.log(complex3.complexAC, complex4.complexAC, similarityScore2nd);
+            if (!complexesOrderedSet.has(complex4)) {
+              complexesOrderedSet.add(complex4);
+            }
+          }
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:88]: Here we could add
```
if (complexesOrderedSet.size === this.complexSearch.elements.length) {
          // All complexes have been added, we can return and stop the loops
          return complexesOrderedSet;
        }
```
To return when we already have all the complexes, so we don't need to keep iterating.

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 19]
```diff
@@ -132,4 +136,59 @@ export class TableStructureComponent implements OnChanges {
...
+    const complexesOrderedSet = this.uniqueComplexesListOrderedBySimilarity(comparedComplexes);
+    // to be used in the table as a 1D array
+    for (const complex of complexesOrderedSet) {
+      finalList.push(complex);
+    }
+    return finalList;
+  }
+
+  uniqueComplexesListOrderedBySimilarity(complexesListSimilarities: [Element, Element, number][]) {
+    const complexesOrderedSet = new Set<Element>();
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:78]: Is `Set` sorted? Set is definitely unique, being a set, but it is not necessarily sorted. What we want is to iterate through the elements in this set (line 165) in the order they were inserted into the set. Can you check if that is happening?

Comment (susiehgt) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:78]: comparedComplexes is sorting the comparedComplexes Array by the highest similarity score before creating the complexesOrderedSet using uniqueComplexesListOrderedBySimilarity,  isn't it enough?

Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:78]: We sort `complexesListSimilarities` before this method, so when we do `for (const [complex1, complex2, similarityScore] of complexesListSimilarities)` we are iterating through them in order.

Based on that, we insert into `comparedComplexes`, also in order. But `comparedComplexes` is a Set, and a set guarantees objects are unique and quick to access, but it does not guarantee it's in order of insertion.

So, when we do `for (const complex of complexesOrderedSet) {` (line 165), we are not guaranteed to iterate through the complexes in order.

We have 2 options:

1. Find an implementation of set that it is ordered. If we find one, we also don't need checks like the following, as set already doesn't allow duplicates.
    ```
    if (!complexesOrderedSet.has(complex1)) {
      complexesOrderedSet.add(complex1);
    }
    ```

2. Use a list. In line 172, we can do `const complexesOrderedSet: Element[] = [];` and it would work without any other changes. If we do this, we also don't need `finalList`, as we already have a list.

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts [hunk 20]
```diff
@@ -204,59 +209,105 @@ export class TableInteractorColumnComponent implements OnChanges {
...
   }
 
   public classifyInteractorsByOccurrence() {
+    this.enrichedInteractors.sort((a, b) =>
+      b.timesAppearing - a.timesAppearing
+    );
+    this.ranges = [];
+  }
+
+  private calculateTimesAppearingType() {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts]: A couple of improvements on these methods. First, we could merge all 3 into a single method, to only iterate through interactors and components once. Second, we could simplify the has/get/set methods to deal with the map.

Regarding the maps
```
if (map.has(key)) {
    map.set(key, map.get(key) + 1);
} else {
    map.set(key, 1);
}
```
Is the same as
```
map.set((key, map.get(key) || 0) + 1);
```

Regarding merging the 3 methods (calculateTimesAppearingType, calculateTimesAppearingOrganism and interactorsTimesAppearing) into one, we could have
```
  private calculateTimesAppearing() {
    // Initialise times appearing by type or organism
    this._timesAppearingByType = new Map();
    this._timesAppearingByOrganism = new Map();
    for (const oneInteractor of this.enrichedInteractors) {
      // Initialise times appearing for each interactor
      oneInteractor.timesAppearing = 0;
      for (const complex of this.complexes) {
        const match = findInteractorInComplex(complex, oneInteractor.interactor.identifier, this.enrichedInteractors);
        if (!!match) {
          // Update times appearing for the interactor
          oneInteractor.timesAppearing += 1;
          // Update times appearing for the interactor type
          this._timesAppearingByType.set(
            oneInteractor.interactor.interactorType,
            (this._timesAppearingByType.get(oneInteractor.interactor.interactorType) || 0) + 1);
          // Update times appearing for the interactor organism
          this._timesAppearingByOrganism.set(
            oneInteractor.interactor.organismName,
            (this._timesAppearingByOrganism.get(oneInteractor.interactor.organismName) || 0) + 1);
        }
      }
    }
  }
```

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts (hunk 19 above)
Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:78]: I had a look at this, and it looks like Set returns the items ordered by insertion, in which case, using a set is fine. Just a few changes.

1. We don't need to call `has`, so instead of
   ```
   if (!complexesOrderedSet.has(complex1)) {
      complexesOrderedSet.add(complex1);
   }
   ```
   You can just do
   ```
   complexesOrderedSet.add(complex1);
   ```

2. Simply making the set into a list. Instead of
   ```
    for (const complex of complexesOrderedSet) {
      finalList.push(complex);
    }
    return finalList;
   ```
   You can do
   ```
   return Array.from(complexesOrderedSet);
   ```

3. Remove the commented out code.

File: src/app/basket/basket.component.ts [hunk 21]
```diff
@@ -45,4 +54,50 @@ export class BasketComponent implements OnInit, AfterViewInit {
...
+  private createQuery(object: any): string {
+    let query = '';
+    for (const key of this.getKeys(object)) {
+      if (object[key] !== undefined) {
+        let queryPerObject = ' complex_id:';
+        queryPerObject += object[key].id;
+        query += queryPerObject;
+      }
+    }
+    return query;
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.ts]: This method can be simplified to just one line
```
return 'complex_id:' + Object.values(object).map((v: BasketItem) => v.id).join(',');
```

File: src/app/basket/basket.component.ts [hunk 22]
```diff
@@ -45,4 +54,50 @@ export class BasketComponent implements OnInit, AfterViewInit {
...
+  }
+
+  private requestComplexesForNavigator() {
+    this.complexPortalService.findComplex(this.createQuery(this._complexBasket)).subscribe(complexSearch => {
```
Comment (jmedinaebi) [inline src/app/basket/basket.component.ts]: To fix the issue with only displaying 10 complexes in the basket, we need to set the the page size to the right value. By default it is 10.
```
const pageSize = Object.values(this._complexBasket).length;
    this.complexPortalService.findComplex(this.createQuery(this._complexBasket), [], [], [], 1, pageSize)
      .subscribe(complexSearch => {
```

File: src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts [hunk 23]
```diff
@@ -327,4 +342,44 @@ export class TableInteractorColumnComponent implements OnChanges {
     }
     return null;
   }
+
+  interactorsTimesAppearing() {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts]: This method is no longer needed, we already increase `oneInteractor.timesAppearing` in line 372.

File: src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts [hunk 24]
```diff
@@ -19,117 +20,128 @@ export class TableStructureComponent implements OnChanges {
...
-          }
-        }
+        components.push(component);
       }
     }
-    return similarities;
+    return components;
+  }
+
+  private calculateSimilarity(complex1: Element, complex2: Element) {
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-navigator/table-structure/table-structure.component.ts:49]: I like it, simple and short.

**Comment by jmedinaebi** [comment]
The style still looks a bit weird.

This is how the list view and navigator view looks in https://www.ebi.ac.uk/complexportal/complex/search?query=%22nuclear%20pore%22&page=1#view_list

Screenshot 2024-06-25 at 09 12 36

Screenshot 2024-06-25 at 09 12 29

And this is how it currently looks in the github pages https://complex-portal.github.io/complex-portal-view/complex/search?query=%22nuclear%20pore%22&page=1#view_list

Screenshot 2024-06-25 at 09 12 43

Screenshot 2024-06-25 at 09 12 49

I know we needed some changes for Firefox and Safari, but it looked better before. Can we revert some of the changes to go back closer to how it used to look? Particularly the padding on the left of the page.

File: src/app/complex/complex-results/complex-results.component.css [hunk 25]
```diff
@@ -36,12 +36,12 @@
...
 
 .filterColumn {
-  width: calc(100rem * 0.22);
+  width: 15vw;
 }
 
 .leftAligned {
   --global-margin: calc((100vw - 80rem) / -2);
-  --min-margin-left: calc(-100rem * 0.17);
+  --min-margin-left: calc(-100rem * 1.17);
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-results.component.css]: Try changing this back to how it was `--min-margin-left: calc(-100rem * 0.17);` to align the filters properly.

File: src/app/complex/complex-results/complex-results.component.css [hunk 26]
```diff
@@ -51,30 +51,42 @@
...
-.cnButton.selected {
-  animation: none;
+.column:last-child:not(:first-child),
+.columns:last-child:not(:first-child) {
+  float: left !important;
+}
+
+.listOfResults {
+  width: 40vw;
+  margin-left: 17vw;
```
Comment (jmedinaebi) [inline src/app/complex/complex-results/complex-results.component.css]: If you change this to 10vw, it may look a bit better `margin-left: 17vw;`

Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-results.component.css]: After this and the other CSS change mentioned above, try in Chrome, Safari and Firefox to check everything looks ok.

File: src/app/complex/complex-results/complex-results.component.css (hunk 25 above)
Comment (jmedinaebi) [reply src/app/complex/complex-results/complex-results.component.css]: Maybe it doesn't need to change to 1.17 again, but something in the middle. At the moment it's too close to the left border, almost as if there is no margin at all.

**Comment by jmedinaebi** [comment]
Another issue, when on a big screen
Screenshot 2024-06-25 at 09 12 43

It does not happen if I have the window on my laptop screen, but on the big screen it does.
